controller/volume_controller.go

```diff
@@ -1712,9 +1708,12 @@ func (c *VolumeController) openVolumeDependentResources(v *longhorn.Volume, e *l
 		if r.Status.CurrentState == longhorn.InstanceStateError {
 			continue
 		}
-		// wait for all potentially healthy replicas become running
-		if r.Status.CurrentState != longhorn.InstanceStateRunning {
-			return nil
+		// wait at least one replica is running
```

Comment:
@shuo-wu @derekbit @PhanLe1010 WDYT?

Reply:
I think this may be dangerous. If all replicas are failed, we may auto-salvage with only one replica. We set `r.Spec.FailedAt = ""` for all replicas we want to autosalvage [here](https://github.com/longhorn/longhorn-manager/blob/6b7d12b397738a6ae759f2414e53f7f1b4ba8f30/controller/volume_controller.go#L1305). [Later](https://github.com/longhorn/longhorn-manager/blob/6b7d12b397738a6ae759f2414e53f7f1b4ba8f30/controller/volume_controller.go#L1353) we call `reconcileAttachDetachStateMachine` which [calls](https://github.com/longhorn/longhorn-manager/blob/6b7d12b397738a6ae759f2414e53f7f1b4ba8f30/controller/volume_controller.go#L1433) `openVolumeDependentResources`, where this change lives. If the most up-to-date replica isn't running, we will auto-salvage without it and lose data. WDYT?

Reply:
Good point. I think you could be right.

Q: When auto salvage happens, how are replicas getting rebuilt? Will the volume start with the salvaged replicas first? after those ready, replenish the remaining unscheduled replicas? If yes, the concern still exists?

Reply:
@Vicente-Cheng, it looks like you are referencing [this line of code](https://github.com/longhorn/longhorn-manager/blob/6b7d12b397738a6ae759f2414e53f7f1b4ba8f30/controller/volume_controller.go#L1291), which we use to determine when the last replica failed. However, there is no break from the loop. We continue to check other replicas and add all of them to `failedUsableReplicas`.

[Later](https://github.com/longhorn/longhorn-manager/blob/6b7d12b397738a6ae759f2414e53f7f1b4ba8f30/controller/volume_controller.go#L1303-L1311), we set `r.spec.FailedAt = ""` for ALL replicas that have failed within AutoSalvageTimeLimit (1 minute).

```golang
// Bring up the replicas for auto-salvage
for _, r := range failedUsableReplicas {
	if util.TimestampWithinLimit(lastFailedAt, r.Spec.FailedAt, AutoSalvageTimeLimit) {
		r.Spec.FailedAt = ""
		log.WithField("replica", r.Name).Warn("Automatically salvaging volume replica")
		msg := fmt.Sprintf("Replica %v of volume %v will be automatically salvaged", r.Name, v.Name)
		c.eventRecorder.Event(v, corev1.EventTypeWarning, constant.EventReasonAutoSalvaged, msg)
		salvaged = true
	}
}
```

If I understand correctly, we start the engine with all qualifying replicas, and then let the engine [decide ](https://github.com/longhorn/longhorn-engine/blob/e05a119a9cfab1928b138e036cd1dcd0e97eaf91/pkg/controller/control.go#L580-L610) which replica to use as the source of truth based on `revisionCounter`. Longhorn-manager narrows down the list to only those replicas that failed within a minute of each other, but longhorn-engine makes the final call:

```golang
// checkReplicasRevisionCounter will check if any replica has unmatched
// revision counter, and mark unmatched replica as 'ERR' state.
func (c *Controller) checkReplicasRevisionCounter() error {
	var expectedRevision int64

	revisionCounters := make(map[string]int64)
	for _, r := range c.replicas {
		// The related backend is nil if the mode is ERR
		if r.Mode == types.ERR {
			continue
		}
		counter, err := c.backend.GetRevisionCounter(r.Address)
		if err != nil {
			return err
		}
		if counter > expectedRevision {
			expectedRevision = counter
		}
		revisionCounters[r.Address] = counter
	}

	for address, counter := range revisionCounters {
		if counter != expectedRevision {
			logrus.Errorf("Revision conflict detected! Expect %v, got %v in replica %v. Mark as ERR",
				expectedRevision, counter, address)
			c.setReplicaModeNoLock(address, types.ERR)
		}
	}

	return nil
}

If we start the engine with only one replica, there may be another with a higher revision counter. Do you see it differently?
```

Reply:
Hi @ejweber, great analyst. You are correct about it!
I do not take care about the case that all replicas are failed in one minute. Thanks for catching up!

Yes, with the whole code flow.  If all replicas fail in one minute, they will all be the salvage candidates.
Then if we are just waiting for one replica, we cannot use the latest revision to recover.

The original issue I tried to introduce this change for the first-time creation because we sometimes got stuck on the engine with one of the replicas not ready, and the volume could not be active.

Could we still improve the creation and ensure the auto-salvage stage will wait for all replicas?
Like add checking with `e.Spec.SalvageRequested`? WDUT?
@ejweber @innobead @derekbit @shuo-wu @PhanLe1010

Reply:
Seems reasonable to me. Curious what others think. Thanks for looking into this!

Reply:
The issue here is not only about the salvage part. It's more about the attachment flow instead.
Starting the engine with at least one replica means that:
1. Other healthy replicas may not connect to the engine when the engine is up and handling IO. Then those replicas should be considered as staled and Longhorn needs to pay more resources to do rebuilding. Besides, skipping waiting for all healthy replicas means the HA is hurt/not guaranteed at the beginning.
2. As Eric mentioned, without the wait, we may miss the correct replica candidates and lead to data loss.

Here, I think the correct way is, if a replica cannot be started and it blocks the volume attachment, we should recognize the case and **mark this kind of replica as `Failed` in time** (after several retries). 
I remembered that there was a ticket for this but I cannot find it. And I am not sure if there is a better way to improve this part

Reply:
Agree that. I removed this commit seems there are various points we need to take care of for when the engine could be set as running. It would be better to discuss and improve after that.

I will open an enhancement issue to record it. Thanks for your opinions!